### PR TITLE
Add type hints to application services

### PR DESCRIPTION
**Pull request by clokep**
This adds more type hints to the AS code.

This also fixes a bug in the feature introduced in #8437.

### CODE REVIEW

File: synapse/handlers/appservice.py [hunk 1]
```diff
@@ -261,14 +267,15 @@ async def _handle_presence(
                     ),
                 }
                 for event in presence_events
-            ]
-            events = events + presence_events
+            )
+
+        return events
```
Comment (clokep) [inline synapse/handlers/appservice.py:272]: @Half-Shot I believe this is an bug in the original code?

Comment (Half-Shot) [reply synapse/handlers/appservice.py:272]: Yeah, ugh.

Comment (clokep) [reply synapse/handlers/appservice.py:307]: This got uplifted to 1.22.0rc2 in #8656.

**Comment by clokep** [comment]
@Half-Shot This might be worth you looking over to see if there's anything surprising. I tried my best to figure out the proper types though.

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**